### DISCUSSION TITLE
Migration docs from Xamarin to .NET 6 Mobile

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#9530] [Q&A]
[_Question from AGR on Discord:_](https://discord.com/channels/372137812037730304/550416151172087808/1009437408707756072)

Hello, is there ANY sort of migration guide regarding transition from .NET 5/Xamarin.iOS/Xamarin.Android towards .NET 6/.NET6-iOS/.NET6-Android? Manually tampering with the project dependencies is giving me an absolute headache, especially on windows with the confusing dependency trees throughout the WindowsSDK, .NET and the CSWinRT. I'd just create a new empty solution targeting .NET 6 and WinUI, but I'd like to maintain the legacy project structure with seperate heads.

If someone happens to have an overview/csproj files that work in this regard it would be greatly appreciated.

**Comment by jeromelaban** [comment]
We do not have documentation on migrating to .NET 6 mobile from Xamarin yet. The only guidance we suggest for now is to create a separate solution, then replace the projects in your solution while [moving to WinAppSDK/WinUI 3 namespaces](https://platform.uno/docs/articles/updating-to-winui3.html?q=winui%203).

Now, your scenario is about maintaining both Xamarin and `net6-*` targets (if I understand that correctly), which is a different approach. You're likely using the UWP/WinUI 2 namespaces, which means that you'll need to create a template using `dotnet new unoapp-uwp-net6` to keep the Xamarin compatibility. You can migrate to WinUI 3 later on.

If you need an example of app that uses a similar two-faceted solution, take a look at the [Uno.Gallery](https://github.com/unoplatform/Uno.Gallery) project, where both Xamarin and "Mobile" heads are present.

**Comment by arg-drummond** [comment]
Thanks for answering, I realize I should have phrased my question less ambiguous:

- I already am using WinUI 3 and the Windows project is targeting the WinSDK &. NET 6
- I'd like to keep the old Uno project structure of having different heads for Android and iOS
- these two projects are still running on Xamarin & .NET 5 as I am struggling to find the correct references and properties I need to replace

The natural approach would be to just create a new project with the .NET 6 template and merge all the files into it, but since MAUI uses a single project approach and Uno combined mobile platforms into one as well the project-files look very different.

Microsoft provides minimal information on how to achieve this sort of migration: to [.NET6-ios](https://github.com/xamarin/xamarin-macios/wiki/Project-file-properties-dotnet-migration), to [.NET6-droid](https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/OneDotNet.md) (for example, I did most of what was listed there for iOS, but whatever becomes of `"<Reference Include="Xamarin.iOS" />"` is never explained). Standalone iOS or Android projects do not seem to be thing in .NET 6 anymore as well so I couldn't find any sort of .csproj-file for this matter either.

I was hoping that maybe Uno has been considering this and made some sort of guidance.

**Reply by jeromelaban** [reply]
Thanks for the details. Here's an [updated version of the migration document](https://github.com/unoplatform/uno/blob/master/doc/articles/migrating-from-xamarin-to-net6.md), yet you're providing points that are loose in the documentation. We'll provide additional guidance.

> for example, I did most of what was listed there for iOS, but whatever becomes of "<Reference Include="Xamarin.iOS" />" is never explained

The list of references to migrate is indeed not explicitly defined, but in general, `<Reference />` nodes cannot be ported over to .NET 6.

> Standalone iOS or Android projects do not seem to be thing in .NET 6 anymore as well so I couldn't find any sort of .csproj-file for this matter either.

Standalone projects are definitely possible, and for that, the uno project structure is more appropriate because of the use of a Shared Project. It's mostly a matter of splitting the "single project" into multiple files, removing targets from `<TargetFrameworks>`, then removing the blocks for which the conditionals do not match the single target.

**Reply by arg-drummond** [reply]
> It's mostly a matter of splitting the "single project" into multiple files, removing targets from <TargetFrameworks>, then removing the blocks for which the conditionals do not match the single target.

 Seems like the obvious way to go! Didn't think of it myself. First attempts show that VS refuses to acknowledge the altered iOS-project as such and does not show the .NET6.0-ios dependency, but rather the old dependencies with waning icons, resulting in failing builds due to missing UIKit namespaces. If it's not a caching issue it looks like I have keep digging.

**Reply by jeromelaban** [reply]
Know that you cannot mix and match `xamarinios` and `net6.0-ios` projects or nuget packages in the same solution. Is that the case in your project?

**Reply by arg-drummond** [reply]
After a lot of trial and error with build the configurations I think I finally got it, it looks like this:

```
<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

	<PropertyGroup>
		<Nullable>enable</Nullable>
		<TargetFrameworks>net6.0-ios</TargetFrameworks>
		<SingleProject>true</SingleProject>
		<OutputType>Exe</OutputType>
		<RuntimeIdentifier>iossimulator-x64</RuntimeIdentifier>	
		<IsUnoHead>true</IsUnoHead>
		<SupportedOSPlatformVersion>15.0</SupportedOSPlatformVersion>

		<EnableDefaultCompileItems>false</EnableDefaultCompileItems>
	</PropertyGroup>

	<PropertyGroup>
		<MtouchExtraArgs>$(MtouchExtraArgs) --setenv=MONO_GC_PARAMS=soft-heap-limit=512m,nursery-size=64m,evacuation-threshold=66,major=marksweep,concurrent-sweep</MtouchExtraArgs>
		<MtouchExtraArgs>$(MtouchExtraArgs) --marshal-objectivec-exceptions:disable</MtouchExtraArgs>
	</PropertyGroup>

	<ItemGroup>
		<Compile Include="Main.cs" />
		<BundleResource Include="Assets\SplashScreen%402x.png" />
		<BundleResource Include="Assets\SplashScreen%403x.png" />
		<BundleResource Include="Assets\Default-568h%402x.png" />
	</ItemGroup>

	<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|iPhone' ">
		<DebugSymbols>true</DebugSymbols>
		<DebugType>full</DebugType>
		<Optimize>false</Optimize>
		<OutputPath>bin\iPhone\Debug</OutputPath>
		<DefineConstants>DEBUG</DefineConstants>
		<ErrorReport>prompt</ErrorReport>
		<WarningLevel>4</WarningLevel>
		<ConsolePause>false</ConsolePause>
		<MtouchArch>ios-arm64</MtouchArch>
		<CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
		<CodesignKey>iPhone Developer</CodesignKey>
		<MtouchDebug>true</MtouchDebug>
		<MtouchExtraArgs>--setenv=MONO_LOG_LEVEL=debug --setenv=MONO_LOG_MASK=gc --setenv=MONO_GC_PARAMS=soft-heap-limit=512m,nursery-size=64m,evacuation-threshold=66,major=marksweep,concurrent-sweep</MtouchExtraArgs>
		<CodesignProvision>Native Debug iPhone</CodesignProvision>
		<PlatformTarget>x64</PlatformTarget>
		<MtouchProfiling>false</MtouchProfiling>
		<DeviceSpecificBuild>true</DeviceSpecificBuild>
		<MtouchInterpreter>-all</MtouchInterpreter>
		<MtouchFastDev>false</MtouchFastDev>
		<MtouchEnableSGenConc>true</MtouchEnableSGenConc>
	</PropertyGroup>
	<PropertyGroup Condition=" '$(RunConfiguration)' == 'Default' ">
		<AppExtensionDebugBundleId />
	</PropertyGroup>
	<PropertyGroup>
		<AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
	</PropertyGroup>

	<ItemGroup>
		<BundleResource Include="Assets\Fonts\uno-fluentui-assets.ttf" />
	</ItemGroup>

	<ItemGroup>
		<!--icon refs-->
	</ItemGroup>
	<ItemGroup>
		<!--nuget refs-->
	</ItemGroup>
	<ItemGroup>
		<InterfaceDefinition Include="LaunchScreen.storyboard" />
	</ItemGroup>
	<ItemGroup>
		<!--project refs-->
	</ItemGroup>
```
Though two things bother me:

1. The info.plist is expected in a subfolder despite nothing hinting to this in the csproj file.
This causes a build error if I do not set `		<EnableDefaultCompileItems>false</EnableDefaultCompileItems>`. Is there any way to get rid of this?
![image](https://user-images.githubusercontent.com/112620248/188472276-23949408-9b47-4de5-909e-4cd49a488f63.png)

2. I get a runtime error when referencing 'Xamarin.Essentials'. 
```
 Xamarin.Essentials.NotImplementedInReferenceAssemblyException: This functionality is not implemented in the portable version of this assembly. You should reference the NuGet package from your main application project in order to reference the platform-specific implementation.
   at Xamarin.Essentials.DeviceInfo.GetModel() 
```
I am aware that this dependency is replaced by 'Maui.Essentials' but do I really have to migrate all my dependencies to this before I can even run the project without errors? 'Xamarin.Essentials' targets .NET Standard 1.0 after all which should run with .NET 6.
